codec_packetizers/h264/h264_depacketizer.c

```diff
@@ -0,0 +1,304 @@
+
+    if( ( pCtx == NULL ) ||
+        ( pPacketsArray == NULL ) ||
+        ( packetsArrayLength == 0 ) )
+    {
+        result = H264_RESULT_BAD_PARAM;
+    }
+
+    if( result == H264_RESULT_OK )
+    {
+        pCtx->pPacketsArray = pPacketsArray;
+        pCtx->packetsArrayLength = packetsArrayLength;
+
+        pCtx->headIndex = 0;
+        pCtx->tailIndex = 0;
+        pCtx->packetCount = 0;
+    }
+
+    return result;
+}
+
+/*-----------------------------------------------------------*/
+
+H264Result_t H264Depacketizer_AddPacket( H264DepacketizerContext_t * pCtx,
+                                         const H264Packet_t * pPacket )
+{
+    H264Result_t result = H264_RESULT_OK;
+
+    if( ( pCtx == NULL ) ||
+        ( pPacket == NULL ) )
+    {
+        result = H264_RESULT_BAD_PARAM;
+    }
+
+    if( result == H264_RESULT_OK )
+    {
+        if( pCtx->packetCount >= pCtx->packetsArrayLength )
+        {
+            result = H264_RESULT_OUT_OF_MEMORY;
+        }
+    }
+
+    if( result == H264_RESULT_OK )
+    {
+        pCtx->pPacketsArray[ pCtx->headIndex ].pPacketData = pPacket->pPacketData;
+        pCtx->pPacketsArray[ pCtx->headIndex ].packetDataLength = pPacket->packetDataLength;
+        pCtx->headIndex += 1;
+        pCtx->packetCount += 1;
+    }
+
+    return result;
+}
+
+/*-----------------------------------------------------------*/
+
+H264Result_t H264Depacketizer_GetNalu( H264DepacketizerContext_t * pCtx,
+                                       Nalu_t * pNalu )
+{
+    H264Result_t result = H264_RESULT_OK;
+    uint8_t packetType;
+
+    if( ( pCtx == NULL ) ||
+        ( pNalu == NULL ) )
+    {
+        result = H264_RESULT_BAD_PARAM;
+    }
+
+    if( result == H264_RESULT_OK )
+    {
+        if( pCtx->packetCount == 0 )
+        {
+            result = H264_RESULT_NO_MORE_NALUS;
+        }
+    }
+
+    if( result == H264_RESULT_OK )
+    {
+        packetType = ( pCtx->pPacketsArray[ pCtx->tailIndex ].pPacketData[ 0 ] &
+                       NALU_HEADER_TYPE_MASK );
+
+        if( ( packetType >= SINGLE_NALU_PACKET_TYPE_START ) &&
+            ( packetType <= SINGLE_NALU_PACKET_TYPE_END ) )
+        {
+            DepacketizeSingleNaluPacket( pCtx,
+                                         pNalu );
+        }
+        else if( packetType == FU_A_PACKET_TYPE )
+        {
+            result = DepacketizeFragmentationUnitPacket( pCtx,
+                                                         pNalu );
+        }
+        else
+        {
+            result = H264_RESULT_UNSUPPORTED_PACKET;
+        }
+    }
+
+    return result;
+}
+
+/*-----------------------------------------------------------*/
+
+H264Result_t H264Depacketizer_GetFrame( H264DepacketizerContext_t * pCtx,
+                                        Frame_t * pFrame )
+{
+    H264Result_t result = H264_RESULT_OK;
+    Nalu_t nalu;
+    size_t currentFrameDataIndex = 0;
+    uint8_t startCode[] = { 0x00, 0x00, 0x00, 0x01 };
```

Comment:
IIRC, the annex-b nalu can either have 0x00000001 or 0x000001. Is there a reason we are restricting the startCode to one of possibilities only?

Reply:
@disa6302 In create paylaod path we are using both [3byte](https://github.com/moninom1/amazon-kinesis-video-streams-rtp/blob/main/codec_packetizers/h264/h264_packetizer.c#L156C13-L156C23) and [4byte](https://github.com/moninom1/amazon-kinesis-video-streams-rtp/blob/main/codec_packetizers/h264/h264_packetizer.c#L155) code.
This is the depay path and here we are appending only 4byte code which is also inline with webrtc SDK [depayH264FromRtpPayload](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/master/src/source/Rtp/Codecs/RtpH264Payloader.c#L238) Let me know if it seems fine.

Reply:
Got it. Ok. Wanted to confirm. Did not check the SDK implementation for this. Looks good!